src/extensions/logic.rs

```diff
@@ -0,0 +1,113 @@
+//! Basic logical operations.
+
+use std::collections::HashMap;
+
+use itertools::Itertools;
+use smol_str::SmolStr;
+
+use crate::{
+    resource::{OpDef, ResourceSet, SignatureError},
+    types::{
+        type_param::{TypeArg, TypeArgError, TypeParam},
+        SimpleType,
+    },
+    Resource,
+};
+
+/// The resource identifier.
+pub const fn resource_id() -> SmolStr {
+    SmolStr::new_inline("Logic")
+}
+
+/// Construct a boolean type.
+pub fn bool_type() -> SimpleType {
+    SimpleType::new_simple_predicate(2)
+}
+
+/// Resource for basic logical operations.
+pub fn resource() -> Resource {
+    let mut resource = Resource::new(resource_id());
+
+    let not_op = OpDef::new_with_custom_sig(
+        "Not".into(),
+        "logical 'not'".into(),
+        vec![],
+        HashMap::default(),
+        |_arg_values: &[TypeArg]| {
+            Ok((
+                vec![bool_type()].into(),
+                vec![bool_type()].into(),
+                ResourceSet::default(),
+            ))
+        },
+    );
+
+    let and_op = OpDef::new_with_custom_sig(
+        "And".into(),
+        "logical 'and'".into(),
+        vec![TypeParam::Int],
+        HashMap::default(),
+        |arg_values: &[TypeArg]| {
+            let a = arg_values.iter().exactly_one().unwrap();
+            let n: Result<&u128, SignatureError> = match a {
```

Comment:
Why not just `u128` ?

Reply:
> Ah, I meant replacing all the `Result<&u128, SignatureError>` with `u128` since it's immediately unwrapped.
> 
> Not that important though; it gets elided by the compiler.

Ah, whoops.